caravel/views.py

```diff
@@ -945,15 +967,113 @@ def msg(self):
 
 class Caravel(BaseCaravelView):
     """The base views for Caravel!"""
+    @log_this
+    @has_access
+    @expose("/request_access_form/<datasource_type>/<datasource_id>/"
+            "<datasource_name>")
+    def request_access_form(
+            self, datasource_type, datasource_id, datasource_name):
+        request_access_url = (
+            '/caravel/request_access?{}={}&datasource_name={}'.format(
+                datasource_type, datasource_id, datasource_name)
+        )
+        return self.render_template(
+            'caravel/request_access.html',
+            request_access_url=request_access_url,
+            datasource_name=datasource_name,
+            slicemodelview_link='/slicemodelview/list/')
+
+    @log_this
+    @has_access
+    @expose("/request_access")
+    def request_access(self):
+        table_id = request.args.get('table')
+        druid_datasource_id = request.args.get('druid')
+        datasource_name = request.args.get('datasource_name')
+
+        duplicates = (
+            db.session.query(models.DatasourceAccessRequest)
+            .filter(
+                models.DatasourceAccessRequest.created_by_fk ==
+                g.user.get_id(),
+                models.DatasourceAccessRequest.table_id == table_id,
+                models.DatasourceAccessRequest.druid_datasource_id ==
+                druid_datasource_id,
+            )
+        ).all()
+
+        if len(duplicates) > 0:
+            flash(__(
+                "You have already requested access to the datasource %(name)s",
+                name=datasource_name), "warning")
+            return redirect('/slicemodelview/list/')
+
+        access_request = models.DatasourceAccessRequest(
+            table_id=table_id,
+            druid_datasource_id=druid_datasource_id,
+        )
+        db.session.add(access_request)
+        db.session.commit()
+        flash(__("Access to the datasource %(name)s was requested",
+                 name=datasource_name), "info")
+        return redirect('/slicemodelview/list/')
+
+    @log_this
+    @has_access
+    @expose("/approve")
+    def approve(self):
+        request_access_id = request.args.get('request_access_id')
+        role_to_grant = request.args.get('role_to_grant')
+        role_to_extend = request.args.get('role_to_extend')
+        access_request = (
+            db.session.query(models.DatasourceAccessRequest)
+            .filter_by(id=request_access_id).first()
+        )
+        # check if you can approve
+        if (self.all_datasource_access() or
```

Comment:
this endpoint and model related perms should be added to the admin role only (check the cil's init function), then we don't need anything else than `@has_access`.